File: concept/10_array/binary_search/search_circular_sorted/search_circular_sorted.cpp
#include <iostream>
using namespace std;

int getPivot(int arr[], int n)
{

    int start = 0;
    int end = n - 1;
    int mid = start + (end - start) / 2;

    while (start < end)
    {
        if (arr[mid] >= arr[0])
        {
            start = mid + 1;
        }
        else
        {
            end = mid;
        }
        mid = start + (end - start) / 2;
    }
    return start;
}

int binarySearch(int arr[], int start, int end, int key)
{

    // int start = start;
    // int end = end - 1;
    int mid = start + (end - start) / 2;

    while (start <= end)
    {

        if (arr[mid] == key)
        {
            return mid;
        }

        if (key > arr[mid])
        {
            start = mid + 1;
        }
        else
        {
            end = mid - 1;
        }

        mid = start + (end - start) / 2;
    }
    return -1;
}

int main()
{
    int arr[5] = {3, 8, 10, 17, 1};
    int n = 5;
    int key = 17;

    int pivot = getPivot(arr, 5);

    if (key >= arr[pivot] && key <= arr[n - 1])
    {
        int search = binarySearch(arr, pivot, n - 1, key);
        cout << "Pivot point Index is " << pivot << " and value is " << arr[pivot] << endl;
        cout << "Pivot point Index is " << search << endl;
    }
    else
    {
        int search2 = binarySearch(arr, 0, pivot - 1, key);
        cout << "2 Pivot point Index is " << pivot << " and value is " << arr[pivot] << endl;
        cout << "2 Pivot point Index is " << search2 << endl;
    }
}

File: concept/10_array/binary_search/search_circular_sorted/search_circular_sorted.dart
void main() {
  List<int> pivotList = [8, 10, 17, 1, 3];
  int n = pivotList.length;
  int key = 17;

  int pivotIndex = getPivot(pivotList);
  print("Pivot index is $pivotIndex , value is ${pivotList[pivotIndex]} ");

  if (key >= pivotList[pivotIndex] && key <= pivotList.last) {
    int searchIndex = binarySearch(pivotList, pivotIndex, n - 1, key);
    print("Pivot index is $searchIndex , value is ${pivotList[searchIndex]} ");
  } else {
    int searchIndex = binarySearch(pivotList, 0, pivotIndex - 1, key);
    print("Pivot index is $searchIndex , value is ${pivotList[searchIndex]} ");
  }
}

int getPivot(List<int> data) {
  int start = 0;
  int end = data.length - 1;
  int mid = start + (end - start) ~/ 2;

  while (start < end) {
    if (data[mid] >= data[0]) {
      start = mid + 1;
    } else {
      end = mid;
    }
    mid = start + (end - start) ~/ 2;
  }
  return start;
}

int binarySearch(List<int> data, int start, int end, int key) {
  // int start = 0;
  // int end = data.length - 1;
  int median = start + (end - start) ~/ 2;

  while (start <= end) {
    int valueMedian = data[median];
    if (valueMedian == key) {
      return median;
    }

    if (key > valueMedian) {
      start = median + 1;
    } else {
      end = median - 1;
    }

    median = start + (end - start) ~/ 2;
  }

  return -1;
}
